node/network/availability-distribution/src/session_cache.rs

```diff
@@ -134,18 +137,32 @@ impl SessionCache {
 			}
 		};
 
-		if let Some(info) = self.session_info_cache.get(&session_index) {
-			return Ok(Some(with_info(info)));
+		if let Some(o_info) = self.session_info_cache.get(&session_index) {
+			tracing::trace!(target: LOG_TARGET, session_index, "Got session from lru");
+			if let Some(info) = o_info {
+				return Ok(Some(with_info(info)));
+			}
+			else {
+				// Info was cached - we are not a validator: return early:
+				return Ok(None)
+			}
 		}
 
 		if let Some(info) = self
 			.query_info_from_runtime(ctx, parent, session_index)
 			.await?
 		{
+			tracing::trace!(target: LOG_TARGET, session_index, "Calling `with_info`");
 			let r = with_info(&info);
-			self.session_info_cache.put(session_index, info);
+			tracing::trace!(target: LOG_TARGET, session_index, "Storing session info in lru!");
+			self.session_info_cache.put(session_index, Some(info));
 			return Ok(Some(r));
 		}
+		else {
+			// Avoid needless fetches if we are not a validator:
+			self.session_info_cache.put(session_index, None);
```

Comment:
I don't think if `None` is returned it necessary means we're not a validator. What makes you think that's the case?

Reply:
[query_info_from_runtime](https://github.com/paritytech/polkadot/blob/96063da161892c4967084499cdef8a5529faec72/node/network/availability-distribution/src/session_cache.rs#L216) returns `Nothing` if [get_our_index](https://github.com/paritytech/polkadot/blob/96063da161892c4967084499cdef8a5529faec72/node/network/availability-distribution/src/session_cache.rs#L284) is not able to find our own `ValidatorIndex`. It is assumed that this is the case, when we are in fact not a validator.

Reply:
Ah, ok, we return an error if the call for SessionInfo returns None, I see.

> Nothing

good ol' Haskell times ;)

Reply:
> good ol' Haskell times ;)

Haha, you got me. :-)